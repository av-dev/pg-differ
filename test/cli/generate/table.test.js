'use strict';

const fs = require('fs');
const path = require('path');

const helpers = require('../../helpers');
const { connectionString } = require('../../pg.config');

const cliPath = path.join(helpers.rootPath, 'bin', 'cli');
const dictionary = path.join(__dirname, '__schemas__');

describe(`cli generate`, () => {
  const roles = {
    name: 'DifferSchema.roles',
    unique: [{ columns: ['id'] }],
    columns: { id: 'integer' },
  };
  const users = {
    name: 'DifferSchema.users',
    foreignKeys: [
      {
        columns: ['role'],
        references: {
          table: roles.name,
          columns: ['id'],
        },
      },
    ],
    columns: {
      id: { type: 'smallint', identity: true, primary: true },
      birthday: {
        type: 'timestamp without time zone',
        default: { type: 'literal', value: 'now()' },
      },
      joined: 'timestamp without time zone',
      role: 'integer',
    },
  };

  const sequence = {
    name: 'DifferSchema.users_seq',
    start: '1',
    min: '1',
    max: '9999',
  };

  afterAll(() => {
    helpers.rmdirSync(dictionary, { recursive: true });
  });

  let differ;
  beforeAll(() => {
    differ = helpers.getDiffer();
    differ.define('table', roles);
    differ.define('table', users);
    differ.define('sequence', sequence);
    return differ.sync({ force: true });
  });

  it(`table schema`, async function () {
    await helpers.execute(cliPath, [
      'generate',
      `--connection=${connectionString}`,
      `--path=${dictionary}`,
      `--table=${users.name}`,
      `--no-group`,
    ]);

    expect(
      JSON.parse(
        fs.readFileSync(
          path.join(dictionary, `${users.name}.schema.json`),
          'utf-8'
        )
      )
    ).toEqual({ type: 'table', properties: users });
  });

  it(`sequence schema`, async function () {
    await helpers.execute(cliPath, [
      'generate',
      `--connection=${connectionString}`,
      `--path=${dictionary}`,
      `--sequence=${sequence.name}`,
      `--no-group`,
    ]);

    expect(
      JSON.parse(
        fs.readFileSync(
          path.join(dictionary, `${sequence.name}.schema.json`),
          'utf-8'
        )
      )
    ).toMatchObject({
      type: 'sequence',
      properties: { name: sequence.name, max: sequence.max },
    });
  });

  it(`grouping schemas`, async function () {
    await helpers.execute(cliPath, [
      'generate',
      `--connection=${connectionString}`,
      `--path=${dictionary}`,
      `--table=${users.name}`,
      `--group`,
    ]);

    expect(
      JSON.parse(
        fs.readFileSync(
          path.join(dictionary, 'DifferSchema', 'users.schema.json'),
          'utf-8'
        )
      )
    ).toEqual({ type: 'table', properties: users });
  });
});
