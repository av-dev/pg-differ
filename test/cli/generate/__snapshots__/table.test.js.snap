// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli generate grouping schemas 1`] = `
Object {
  "properties": Object {
    "columns": Object {
      "created": Object {
        "default": Object {
          "type": "literal",
          "value": "now()",
        },
        "type": "timestamptz(2)",
      },
      "deleted": "timestamp(2)",
      "id": "integer",
    },
    "name": "DifferSchema.roles",
    "unique": Array [
      Object {
        "columns": Array [
          "id",
        ],
      },
    ],
  },
  "type": "table",
}
`;

exports[`cli generate sequence schema 1`] = `
Object {
  "properties": Object {
    "max": "9999",
    "name": "DifferSchema.users_seq",
  },
  "type": "sequence",
}
`;

exports[`cli generate table schema 1`] = `
Object {
  "roles": Object {
    "properties": Object {
      "columns": Object {
        "created": Object {
          "default": Object {
            "type": "literal",
            "value": "now()",
          },
          "type": "timestamptz(2)",
        },
        "deleted": "timestamp(2)",
        "id": "integer",
      },
      "name": "DifferSchema.roles",
      "unique": Array [
        Object {
          "columns": Array [
            "id",
          ],
        },
      ],
    },
    "type": "table",
  },
  "users": Object {
    "properties": Object {
      "columns": Object {
        "id": Object {
          "identity": true,
          "primary": true,
          "type": "smallint",
        },
        "role": "integer",
      },
      "foreignKeys": Array [
        Object {
          "columns": Array [
            "role",
          ],
          "references": Object {
            "columns": Array [
              "id",
            ],
            "table": "DifferSchema.roles",
          },
        },
      ],
      "name": "DifferSchema.users",
    },
    "type": "table",
  },
}
`;

exports[`cli generate using --no-pretty-types 1`] = `
Object {
  "properties": Object {
    "columns": Object {
      "created": Object {
        "default": Object {
          "type": "literal",
          "value": "now()",
        },
        "type": "timestamp(2) with time zone",
      },
      "deleted": "timestamp(2) without time zone",
      "id": "integer",
    },
    "name": "DifferSchema.roles",
    "unique": Array [
      Object {
        "columns": Array [
          "id",
        ],
      },
    ],
  },
  "type": "table",
}
`;
